File: snake-ai/SnakeBrain.ts
import { Utils } from "snake-game/utils";
import { Direction } from "snake-game/typing";
import { CalcUtils } from "./CalcUtils";
import type { ActivationFunction } from "./CalcUtils";

export type LayerShape = number[];

interface ProvidedWeightAndBias {
  weightArr: number[][][];
  biasesArr: number[][];
}

interface Options {
  inputLength: number;
  layerShapes: LayerShape[];
  hiddenLayerActivationFunction: ActivationFunction;
  providedWeightAndBias?: ProvidedWeightAndBias;
}

export interface ISnakeBrain {
  inputLength: number;
  layerShapes: LayerShape[];
  readonly hiddenLayerActivationFunction: ActivationFunction;
  weightArr: number[][][];
  biasesArr: number[][];
}

export default class SnakeBrain implements ISnakeBrain {
  public static readonly OUTPUT_LAYER_LENGTH = 4;

  public static readonly actionMap = Object.freeze({
    0: Direction.UP,
    1: Direction.DOWN,
    2: Direction.LEFT,
    3: Direction.RIGHT,
  }) satisfies Record<string, Direction>;

  private static readonly MIN_WEIGHT = -1;
  private static readonly MAX_WEIGHT = 1;
  private static readonly MIN_BIAS = -1;
  private static readonly MAX_BIAS = 1;

  public static crossOverNumber(a: number, b: number): number {
    return Math.random() < 0.5 ? a : b;
  }

  public inputLength: number;
  public layerShapes: LayerShape[];
  public readonly hiddenLayerActivationFunction: ActivationFunction;
  public weightArr: number[][][];
  public biasesArr: number[][];

  constructor(options: Options) {
    this.inputLength = options.inputLength;
    this.layerShapes = options.layerShapes;
    this.hiddenLayerActivationFunction = options.hiddenLayerActivationFunction;
    if (!this.validateLayerShapes()) throw new Error("Invalid layer shapes");

    if (options.providedWeightAndBias) {
      this.weightArr = Utils.cloneObject(options.providedWeightAndBias.weightArr);
      this.biasesArr = Utils.cloneObject(options.providedWeightAndBias.biasesArr);
      if (!this.validateWeightAndBias()) throw new Error("Invalid provided weight and bias");
    } else {
      this.weightArr = this.layerShapes.map((x) => this.generateRandomLayerWeight(x));
      this.biasesArr = this.layerShapes.map((x) => this.generateRandomLayerBiases(x));
    }
  }

  public toPlainObject(): ISnakeBrain {
    return {
      inputLength: this.inputLength,
      layerShapes: this.layerShapes,
      hiddenLayerActivationFunction: this.hiddenLayerActivationFunction,
      weightArr: this.weightArr,
      biasesArr: this.biasesArr,
    };
  }

  public validateWeightAndBias(): boolean {
    if (this.weightArr.length === 0 || this.biasesArr.length === 0) return false;
    if (this.weightArr.length !== this.biasesArr.length) return false;
    if (this.weightArr.length !== this.layerShapes.length) return false;
    if (this.weightArr[0][0].length !== this.inputLength) return false;
    if (this.weightArr[this.weightArr.length - 1].length !== SnakeBrain.OUTPUT_LAYER_LENGTH) return false;
    if (!this.weightArr.flat(2).every((x) => x >= SnakeBrain.MIN_WEIGHT && x <= SnakeBrain.MAX_WEIGHT)) return false;
    if (!this.biasesArr.flat(2).every((x) => x >= SnakeBrain.MIN_BIAS && x <= SnakeBrain.MAX_BIAS)) return false;

    for (let i = 1; i < this.weightArr.length; i++) {
      if (this.weightArr[i].length !== this.layerShapes[i][0]) return false;
      if (this.weightArr[i][0].length !== this.layerShapes[i][1]) return false;
    }

    for (let i = 0; i < this.biasesArr.length; i++) {
      if (this.biasesArr[i].length !== this.layerShapes[i][0]) return false;
    }

    return true;
  }

  public compute(input: number[]): Direction {
    const output = CalcUtils.computeMultipleLayer(this.weightArr, input, this.biasesArr, this.hiddenLayerActivationFunction);
    const index = CalcUtils.indexOfMaxValueInArray(output);
    return this.getAction(index);
  }

  public crossOver(a: SnakeBrain, b: SnakeBrain): void {
    for (let layerIndex = 0; layerIndex < this.weightArr.length; layerIndex++) {
      const weight = this.weightArr[layerIndex];
      const biases = this.biasesArr[layerIndex];

      for (let row = 0; row < weight.length; row++) {
        for (let col = 0; col < weight[row].length; col++) {
          weight[row][col] = SnakeBrain.crossOverNumber(a.weightArr[layerIndex][row][col], b.weightArr[layerIndex][row][col]);
        }

        biases[row] = SnakeBrain.crossOverNumber(a.biasesArr[layerIndex][row], b.biasesArr[layerIndex][row]);
      }
    }
  }

  public mutate(mutationRate: number, mutationAmount: number): void {
    for (let layerIndex = 0; layerIndex < this.weightArr.length; layerIndex++) {
      const weight = this.weightArr[layerIndex];
      const biases = this.biasesArr[layerIndex];

      for (let row = 0; row < weight.length; row++) {
        for (let col = 0; col < weight[row].length; col++) {
          if (Utils.randomBool(mutationRate)) {
            const newValue = weight[row][col] + Utils.randomUniform(-mutationAmount, mutationAmount);
            weight[row][col] = CalcUtils.minmax(SnakeBrain.MIN_WEIGHT, newValue, SnakeBrain.MAX_WEIGHT);
          }
        }

        if (Utils.randomBool(mutationRate)) {
          const newValue = biases[row] + Utils.randomUniform(-mutationAmount, mutationAmount);
          biases[row] = CalcUtils.minmax(SnakeBrain.MIN_BIAS, newValue, SnakeBrain.MAX_BIAS);
        }
      }
    }
  }

  public exportBrainToJson(): string {
    return JSON.stringify(this);
  }

  private validateLayerShapes(): boolean {
    const numOfLayer = this.layerShapes.length;

    if (numOfLayer < 1) return false;

    if (this.layerShapes[0][1] !== this.inputLength) return false;

    if (this.layerShapes[numOfLayer - 1][0] !== SnakeBrain.OUTPUT_LAYER_LENGTH) return false;

    for (let i = 1; i < numOfLayer; i++) {
      if (this.layerShapes[i][1] !== this.layerShapes[i - 1][0]) return false;
    }

    return true;
  }

  private generateRandomLayerWeight(layerShape: LayerShape): number[][] {
    const [row, col] = layerShape;
    const template: number[][] = Array(row)
      .fill(null)
      .map(() => Array(col).fill(null));
    return template.map((x) => x.map((_) => Utils.randomUniform(SnakeBrain.MIN_WEIGHT, SnakeBrain.MAX_WEIGHT)));
  }

  private generateRandomLayerBiases(layerShape: LayerShape): number[] {
    const [row, _] = layerShape;
    const template: number[] = Array(row).fill(null);
    return template.map((_) => Utils.randomUniform(SnakeBrain.MIN_BIAS, SnakeBrain.MAX_BIAS));
  }

  private getAction(index: number): Direction {
    return SnakeBrain.actionMap[index];
  }
}

File: kotlin-root/gasnake/src/main/kotlin/com/example/snake/ai/SnakeBrain.kt
package com.example.snake.ai

import com.example.snake.game.Direction
import com.example.snake.game.Utils
import kotlin.random.Random

data class ProvidedWeightsAndBiases(
    val weights: List<List<DoubleArray>>,
    val biases: List<DoubleArray>
)

data class Options(
    val inputLength: Int,
    val layerShapes: List<List<Int>>,
    val hiddenLayerActivationFunction: ActivationFunction,
    val providedWeightsAndBiases: ProvidedWeightsAndBiases?
)

data class SnakeBrainData(
    val inputLength: Int,
    val layerShapes: List<List<Int>>,
    val hiddenLayerActivationFunction: ActivationFunction,
    val weights: List<List<DoubleArray>>,
    val biases: List<DoubleArray>
)


class SnakeBrain(options: Options) {
    companion object {
        const val OUTPUT_LAYER_LENGTH = 4
        const val MIN_WEIGHT = -10.0
        const val MAX_WEIGHT = 10.0
        const val MIN_BIAS = -10.0
        const val MAX_BIAS = 10.0
        const val MIN_RANDOM_WEIGHT = -1.0
        const val MAX_RANDOM_WEIGHT = 1.0
        const val MIN_RANDOM_BIAS = -1.0
        const val MAX_RANDOM_BIAS = 1.0

        fun crossOverNumber(a: Double, b: Double): Double {
            return if (Random.nextBoolean()) a else b
        }
    }

    val inputLength: Int = options.inputLength
    val layerShapes: List<List<Int>> = options.layerShapes
    val hiddenLayerActivationFunction: ActivationFunction = options.hiddenLayerActivationFunction
    var weights: List<List<DoubleArray>>
    var biases: List<DoubleArray>

    init {
        if (!validateLayerShapes()) throw IllegalArgumentException("Invalid layer shapes")

        if (options.providedWeightsAndBiases != null) {
            // todo check immutability problem
            weights = options.providedWeightsAndBiases.weights
            biases = options.providedWeightsAndBiases.biases
            if (!validateWeightsAndBiases()) throw IllegalArgumentException("Invalid provided weight and bias")
        } else {
            weights = layerShapes.map { generateRandomLayerWeight(it) }
            biases = layerShapes.map { generateRandomLayerBias(it) }
        }
    }

    fun toPlainObject(): SnakeBrainData = SnakeBrainData(
        inputLength = inputLength,
        layerShapes = layerShapes,
        hiddenLayerActivationFunction = hiddenLayerActivationFunction,
        weights = weights.map { x -> x.map { y -> y.clone() } },
        biases = biases.map { it.clone() }
    )

    fun validateWeightsAndBiases(): Boolean {
        if (weights.isEmpty() || biases.isEmpty()) return false
        if (weights.size != biases.size) return false
        if (weights.size != layerShapes.size) return false
        if (weights[0][0].size != inputLength) return false
        if (weights.last().size != OUTPUT_LAYER_LENGTH) return false
        if (weights.flatten().any { arr -> arr.any { it < MIN_WEIGHT || it > MAX_WEIGHT } }) return false
        if (biases.any { arr -> arr.any { it < MIN_BIAS || it > MAX_BIAS } }) return false

        for (i in 1 until weights.size) {
            if (weights[i].size != layerShapes[i][0]) return false
            if (weights[i][0].size != layerShapes[i][1]) return false
        }

        for (i in biases.indices) {
            if (biases[i].size != layerShapes[i][0]) return false
        }

        return true
    }

    fun compute(input: DoubleArray): Direction {
        val output = CalcUtils.computeMultipleLayer(weights, input, biases, hiddenLayerActivationFunction)
        val index = CalcUtils.indexOfMaxValueInArray(output)
        return Direction.inverseDirectionMap(index)
    }

    fun crossOver(a: SnakeBrain, b: SnakeBrain) {
        for (layerIndex in weights.indices) {
            val weight = weights[layerIndex]
            val bias = biases[layerIndex]

            for (row in weight.indices) {
                for (col in weight[row].indices) {
                    weight[row][col] = crossOverNumber(a.weights[layerIndex][row][col], b.weights[layerIndex][row][col])
                }

                bias[row] = crossOverNumber(a.biases[layerIndex][row], b.biases[layerIndex][row])
            }
        }
    }

    fun mutate(mutationRate: Double, mutationAmount: Double) {
        for (layerIndex in weights.indices) {
            val weight = weights[layerIndex]
            val bias = biases[layerIndex]

            for (row in weight.indices) {
                for (col in weight[row].indices) {
                    if (Utils.randomBool(mutationRate)) {
                        val newValue = weight[row][col] + Utils.randomUniform(-mutationAmount, mutationAmount)
                        weight[row][col] = CalcUtils.minmax(MIN_WEIGHT, newValue, MAX_WEIGHT)
                    }
                }

                if (Utils.randomBool(mutationRate)) {
                    val newValue = bias[row] + Utils.randomUniform(-mutationAmount, mutationAmount)
                    bias[row] = CalcUtils.minmax(MIN_BIAS, newValue, MAX_BIAS)
                }
            }
        }
    }

    private fun validateLayerShapes(): Boolean {
        val numOfLayer = layerShapes.size

        if (numOfLayer < 1) return false

        if (layerShapes[0][1] != inputLength) return false

        if (layerShapes[numOfLayer - 1][0] != OUTPUT_LAYER_LENGTH) return false

        for (i in 1 until numOfLayer) {
            if (layerShapes[i][1] != layerShapes[i - 1][0]) return false
        }

        return true
    }

    private fun generateRandomLayerWeight(layerShape: List<Int>): List<DoubleArray> {
        val (row, col) = layerShape
        return List(row) { DoubleArray(col) { Random.nextDouble(MIN_RANDOM_WEIGHT, MAX_RANDOM_WEIGHT) } }
    }

    private fun generateRandomLayerBias(layerShape: List<Int>): DoubleArray {
        val (row, _) = layerShape
        return DoubleArray(row) { Random.nextDouble(MIN_RANDOM_BIAS, MAX_RANDOM_BIAS) }
    }
}